### ISSUE TITLE
MySQL to Oracle Incremental Behavior

### ISSUE THREAD

**Issue by MichalisDBA** [osalvador/ReplicaDB#118]
Hello.

I think incremental mode, although it is working, it has a wrong implementation or i think about it different.

When you use incremental mode from a source database table to a sink database table, i expect only the new rows based on Primary Key to be inserted from source table to sink table.

The problem is replicadb on MERGE uses WHEN MATCHED THEN UPDATE SET.

We have a scenario that we want to transfer data from a mysql database to an oracle database. On oracle database sink table we edit some rows after the replicadb sync. When we run again the replicadb incremental config, it updates and overwrites these previsouly edited rows and this is not what we wanted. We want only the new rows to be inserted and NOT update the existing data.

It would be a good option to have a config parameter to skip and exclude the WHEN MATCHED THEN UPDATE SET and use only WHEN NOT MATCHED THEN INSERT on this situation.

Also after the incremental sync is complete it would be best to truncate the staging table and not having two tables with the same amount of data taking space in database.

Thank you.

**Comment by osalvador** [comment]
Hi @MichalisDBA 

It is true that the `incremental` mode can be a bit confusing. The incremental mode aims to synchronize the source table and the sink table only with the new data added or updated in the source table. 

You want to replicate the new data from the source and append-only to the sink table. If you only want to add new or updated values, a PK would not be necessary in the sink table.

To perform an append-only replication, you must use the `complete` or `complete-atomic` modes along with the `sink.disable.truncate=true` property to disable the `TRUNCATE` statement on the sink table.  For example: 

```properties
######################## ReplicadB General Options ########################
mode=complete
jobs=1
fetch.size=100
verbose=false
############################# Soruce Options ##############################
source.connect=jdbc:mysql://some-mysql:4333/user
source.user=user
source.password=<user password>
source.table=source_table
source.where=last_updated>${lastUpdated}
############################# Sink Options ################################
sink.connect=jdbc:oracle:thin:@some-oracle:1521:user
sink.user=user
sink.password=<user password>
sink.table=sink-table
sink.disable.truncate=true
```

Does this solve your problem?

Finally, the staging table is always deleted after each replication, unless you have set a specific table in the `sink.staging.table` property, in which case it is true that ReplicaDB should truncate it. I'll review this. 

Thank you for reporting!

**Comment by MichalisDBA** [comment]
@osalvador 

It does not solve the problem.

With complete mode i set sink.disable.truncate=true.

If the sink table has a Primary Key and you try to run the config i get:

` ERROR ReplicaDB:136 Got exception running ReplicaDB:
java.util.concurrent.ExecutionException: java.sql.BatchUpdateException: ORA-00001: unique constraint (APPS.SYS_C004381112) violated
`
Complete mode does not append the new data only.

If i delete the Primary Key in the sink table and have it as plain column, then the data get duplicated + the new entries.

Finally yes i had set a custom sink.staging.table that's why at the end it didn't get truncated.

**Comment by MichalisDBA** [comment]
Edit:

For append i have to specify a where clause as i understand from your demo config. The thing is this will increase the complexity because for evey sync we have to somehow get the last_updated value from sink.

I think by having an option to ommit the WHEN MATCHED THEN would be great. 

In my case this would not work nonetheless because i am using it in Oracle 9i and Oracle 9i does not have the enhancement to ignore the WHEN MATCHED THEN UPDATE

http://www.oracle-developer.net/display.php?id=310

`In Oracle 10g, either the INSERT or UPDATE operations are optional (you must have one, but not necessarily both, unlike in 9i). This is a critical enhancement and one which makes the MERGE statement much more flexible.`

So anyway consider it for future release to have an option when using increment mode to oracle sink to skip or not the UPDATE or INSERT operation.

**Comment by MichalisDBA** [comment]
Edit 2:

Another thing i thought for my use case to work with MERGE on incremental mode is for example to modify the

`WHEN MATCHED THEN UPDATE SET trg.NAME = src.NAME`

to

`WHEN MATCHED THEN UPDATE SET trg.NAME = trg.NAME`

This way first of all it works with Oracle 9i and second it appends only the new data to sink table because the previsouly data are updated with the same values of the existing sink table.

I think i have to fork it and somehow try to modify it but my JAVA knowledge is non existent.

**Comment by osalvador** [comment]
Hi @MichalisDBA 

I am not quite understanding your need. In incremental replication the goal is to reduce the number of transferred rows. To do this you have to filter the source data with a key (incremental id or date), but ReplicaDB, for the moment, does not save these keys for the next replication (it requires ReplicaDB to save information). 

https://osalvador.github.io/ReplicaDB/docs/docs.html#incremental 

> Currently, you must store the last value of the column used to determine changes in the source table. In future versions, ReplicaDB will do this automatically.

If you are using incremental replication without filtering the source records, then you are transferring the source table to the sink table completely, and then doing a merge. If this is your case, a complete replication  mode would work. 

In order to be able to reproduce the environment, please provide me with the following data: 

**To Reproduce**
Steps to reproduce the behaviour:
1. Source table DDL
2. Sink table DDL
3. ReplicaDB configuration `options-file`. 

Regards!

**Comment by MichalisDBA** [comment]
@osalvador 

My use case is from a source table transfer the data to a sink table by appending only the new data without using the where clause.

Let's say we have a source table

```
SourceTable
------------
ID, NAME
1, aaa
2, bbb
3, ccc
```

We make an initial incremental sync to the sink table

```
SinkTable
------------
ID, NAME
1, aaa
2, bbb
3, ccc
```

We update some values from SinkTable manually

```
SinkTable
------------
ID, NAME
1, aaa
2, XXX
3, ccc
```

Now the source table has new rows

```
SourceTable
------------
ID, NAME
1, aaa
2, bbb
3, ccc
4, ddd
5, eee
```

We want to append only the new rows on sink table without using the where clause. If we use incremental mode it uses the MERGE function BUT because it has the `WHEN MATCHED THEN UPDATE SET` the sink table would look like this after the next sync

```
SinkTable
------------
ID, NAME
1, aaa
2, bbb
3, ccc
4, ddd
5, eee
```

BUT we wanted to keep the previously edited values. The expected result we wanted is this

```
SinkTable
------------
ID, NAME
1, aaa
2, XXX
3, ccc
4, ddd
5, eee
```

So what i am trying to say is that if there was an option to omit the `WHEN MATCHED THEN UPDATE SET` or `WHEN NOT MATCHED THEN INSERT` then that would be great and the final result would look like the one we wanted. BUT for our use case because we use Oracle 9i it does not have this functionality. You have to specify both `WHEN MATCHED THEN UPDATE SET` and `WHEN NOT MATCHED THEN INSERT` to make the statement work.

So i thought a workaround that if i edit the code and replace

`sql.append(" trg.").append(colName).append(" = src.").append(colName).append(" ,");`

to this

`sql.append(" trg.").append(colName).append(" = trg.").append(colName).append(" ,");`

my use case would work fine.

By using COMPLETE mode with 

`sink.disable.truncate=true`

as you said the final result of the sink table would look like this

```
SinkTable
------------
ID, NAME
1, aaa
2, bbb
3, ccc
1, aaa
2, bbb
3, ccc
4, ddd
5, eee
```

If i had a Primary Key on that sink table and use `sink.disable.truncate=true` then it would throw an exception

`ERROR ReplicaDB:136 Got exception running ReplicaDB: java.util.concurrent.ExecutionException: java.sql.BatchUpdateException: ORA-00001: unique constraint (APPS.SYS_C004381112) violated`

So to summarize. If we want to append only the new data then i have to use the WHERE clause and do some additional work to get the last ID from the sink table before running the sync.

But there is a possibility to do it without WHERE clause using incremental mode and having the option to omit `WHEN MATCHED THEN UPDATE SET` or `WHEN NOT MATCHED THEN INSERT` for Oracle version >= 10

**Comment by MichalisDBA** [comment]
Is there a guide on how to compile the project if i fork it?

**Comment by osalvador** [comment]
Hello @MichalisDBA ,

OK, I understand your need. I think the best strategy is to add to ReplicaDB the support to save the last replication key from the source. 

I can make a feature just for Oracle at the moment, with a feature-flag and that will solve your problem.

Modifying the MERGE statement depending on the Oracle version seems risky, and also not necessary if we save the replication key.

**Comment by osalvador** [comment]
> Is there a guide on how to compile the project if i fork it?

In the pipeline you can review the maven compilation command: https://github.com/osalvador/ReplicaDB/blob/master/.github/workflows/CI_Release.yml#L37

**Comment by MichalisDBA** [comment]
OK @osalvador thanks for the info.

I'll try to build a forked project with our modification for our use case and in the future when the implementation of saving the last replication key is done, i'll use this release.

Thank you for your help.
